### DISCUSSION TITLE
Is there a "caps lock enabled"/"caps lock disabled" event?

### DISCUSSION THREAD

**Discussion by miversen33** [wezterm/wezterm#3441] [Q&A]
I am looking to add an indicator to my statusbar to show whether caps lock is enabled or not, but I don't really see much in the events to show this. Would I have to make some quasi garbage in the key table for this.

Realistically, I am looking for something like [window:leader_is_active()](https://wezfurlong.org/wezterm/config/lua/window/leader_is_active.html) and I suppose ideally for any of the mod keys?

**Comment by wez** [comment]
There isn't a thing for this today. I suppose it would be possible to expose the most recently received set of Modifiers in a given window to facilitate this, but it would need a bit of work:

* Would need to provide a way to operate on/reason about the modifier mask in lua. Currently modifiers are mapped as pipe-separated-strings for use in key assignments.
* Would need to find an efficient way to trigger eg: status/titlebar updates when modifiers change. Those updates are relatively expensive and eg: SHIFT changing state can happen frequently. There's a risk of slowing down all keyboard input to support this sort of feature.  Perhaps there should be an additional config option that is itself a mask of modifiers for which status updates are generated?

**Reply by miversen33** [reply]
That makes sense. Should I create an issue for this? 

> Would need to find an efficient way to trigger

Alternatively, it might be more efficient/better if we could simply query the current key state? Something like
```lua
wezterm.window:is_key_pressed('CAPS')
```

Of course, for a stateful button like "caps lock", that doesn't really work, though really capslock is kind of a special button anyway in that regard. Though again, I don't know how expensive throw away state queries are to perform, so that idea might not be as useful as it sounds to me.

**Reply by wez** [reply]
Yes, please file an enhancement issue for this.

The concern around efficiency is this: assuming that you want to show this status in your status area, the default refresh interval for that is once per second.  There are some events that trigger a more immediate update.  If you wanted the modifiers to be reflected with low latency then you might want them to trigger this update more frequently.  If you are doing a lot in your update-status event handler, then there is a risk of slowing down keyboard input.  There's no fear of that if we don't make it trigger an immediate update, but you need to be prepared for the latency to be potentially very noticeable.
